dashboard/app/jobs/evaluate_rubric_job.rb

```diff
@@ -228,25 +247,75 @@ def self.log_metrics(response)
     )
   end
 
-  RETRIES_ON_TIMEOUT = 2
+  ATTEMPTS_ON_TIMEOUT_ERROR = 2
 
   # Retry just once on a timeout. It is likely to timeout again.
-  retry_on Net::ReadTimeout, Timeout::Error, wait: 10.seconds, attempts: RETRIES_ON_TIMEOUT do
+  retry_on Net::ReadTimeout, Timeout::Error, wait: 10.seconds, attempts: ATTEMPTS_ON_TIMEOUT_ERROR do |job, error|
+    log_metric(metric_name: :TimeoutError)
+    log_to_firehose(job: job, error: error, event_name: 'timeout-error')
+  end
+
+  ATTEMPTS_ON_SERVICE_UNAVAILABLE = 3
+
+  # Retry on a 503 Service Unavailable error, including those returned by aiproxy
+  # when openai returns 500.
+  retry_on ServiceUnavailableError, wait: :exponentially_longer, attempts: ATTEMPTS_ON_SERVICE_UNAVAILABLE do |job, error|
+    agent = error.message.downcase.include?('openai') ? 'openai' : 'none'
+    log_metric(metric_name: :ServiceUnavailable, agent: agent)
+    log_to_firehose(job: job, error: error, event_name: 'service-unavailable', agent: agent)
+  end
+
+  ATTEMPTS_ON_GATEWAY_TIMEOUT = 3
+
+  # Retry on a 504 Gateway Timeout error, including those returned by aiproxy
+  # when openai request times out.
+  retry_on GatewayTimeoutError, wait: :exponentially_longer, attempts: ATTEMPTS_ON_GATEWAY_TIMEOUT do |job, error|
+    agent = error.message.downcase.include?('openai') ? 'openai' : 'none'
+    log_metric(metric_name: :GatewayTimeout, agent: agent)
+    log_to_firehose(job: job, error: error, event_name: 'gateway-timeout', agent: agent)
+  end
+
+  def self.log_metric(metric_name:, agent: nil, value: 1)
     Cdo::Metrics.push(
       AI_RUBRIC_METRICS_NAMESPACE,
       [
         {
-          metric_name: :RetryOnTimeout,
-          value: 1,
+          metric_name: metric_name,
+          value: value,
           dimensions: [
             {name: 'Environment', value: CDO.rack_env},
+            {name: 'Agent', value: agent},
           ],
           unit: 'Count'
         }
       ]
     )
   end
 
+  def self.log_to_firehose(job:, error:, event_name:, agent: nil)
+    options = job.arguments.first
+    script_level = ScriptLevel.find(options[:script_level_id])
+
+    FirehoseClient.instance.put_record(
+      :analysis,
+      {
+        study: AI_RUBRICS_FIREHOSE_STUDY,
+        study_group: 'v0',
+        event: event_name,
+        data_string: "#{error.class.name}: #{error.message}",
+        data_json: {
+          user_id: options[:user_id],
+          requester_id: options[:requester_id],
+          script_level_id: options[:script_level_id],
+          script_name: script_level.script.name,
+          lesson_number: script_level.lesson.relative_position,
+          level_name: script_level.level.name,
+          agent: agent
+        }.to_json
+      }
+    )
+  end
+
   def perform(user_id:, requester_id:, script_level_id:, rubric_ai_evaluation_id: nil)
```

Comment:
personal preference nitpick - The `perform` method is the entrypoint and it's clearer if it's the first method in the class.

Reply:
this could get messy, as methods are also interspersed with directives like `retry_on`. tackling in a follow-up PR

Reply:
* https://github.com/code-dot-org/code-dot-org/pull/58510

I tried to follow https://rubystyle.guide/#consistent-classes, at least the parts of it I agreed with 😅 this is why the directives like `retry_on` still come before the `perform` method.